File: test/test_zcgesv.c
/**
 *
 * @file
 *
 *  PLASMA is a software package provided by:
 *  University of Tennessee, US,
 *  University of Manchester, UK.
 *
 * @precisions mixed zc -> ds
 *
 **/

#include "flops.h"
#include "test.h"
#include "plasma.h"
#include <plasma_core_blas.h>
#include "core_lapack.h"

#include <assert.h>
#include <math.h>
#include <omp.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define COMPLEX

#define A(i_, j_) A[(i_) + (size_t)lda*(j_)]

/***************************************************************************//**
 *
 * @brief Tests ZCPOSV
 *
 * @param[in,out] param - array of parameters
 * @param[in]     run - whether to run test
 *
 * Sets flags in param indicating which parameters are used.
 * If run is true, also runs test and stores output parameters.
 ******************************************************************************/
void test_zcgesv(param_value_t param[], bool run)
{
    //================================================================
    // Mark which parameters are used.
    //================================================================
    param[PARAM_DIM    ].used = PARAM_USE_N;
    param[PARAM_NRHS   ].used = true;
    param[PARAM_PADA   ].used = true;
    param[PARAM_PADB   ].used = true;
    param[PARAM_NB     ].used = true;
    param[PARAM_IB     ].used = true;
    param[PARAM_MTPF   ].used = true;
    param[PARAM_ITERSV ].used = true;
    param[PARAM_ZEROCOL].used = true;
    if (! run)
        return;

    //================================================================
    // Set parameters
    //================================================================
    int n    = param[PARAM_DIM].dim.n;
    int nrhs = param[PARAM_NRHS].i;
    int lda  = imax(1, n + param[PARAM_PADA].i);
    int ldb  = imax(1, n + param[PARAM_PADB].i);
    int ldx  = ldb;
    int ITER;

    int    test = param[PARAM_TEST].c == 'y';
    double tol  = param[PARAM_TOL].d * LAPACKE_dlamch('E');

    //================================================================
    // Set tuning parameters
    //================================================================
    plasma_set(PlasmaTuning, PlasmaDisabled);
    plasma_set(PlasmaNb, param[PARAM_NB].i);
    plasma_set(PlasmaIb, param[PARAM_IB].i);
    plasma_set(PlasmaNumPanelThreads, param[PARAM_MTPF].i);

    //================================================================
    // Allocate and initialize arrays
    //================================================================
    plasma_complex64_t *A = (plasma_complex64_t *)malloc(
        (size_t)lda*n*sizeof(plasma_complex64_t));
    assert(A != NULL);

    int *ipiv = (int*)malloc((size_t)n*sizeof(int));
    assert(ipiv != NULL);

    plasma_complex64_t *B = (plasma_complex64_t *)malloc(
        (size_t)ldb*nrhs*sizeof(plasma_complex64_t));
    assert(B != NULL);

    plasma_complex64_t *X = (plasma_complex64_t *)malloc(
        (size_t)ldx*nrhs*sizeof(plasma_complex64_t));
    assert(X != NULL);

    // Initialize random A
    int seed[] = {0, 0, 0, 1};
    lapack_int retval;
    retval = LAPACKE_zlarnv(1, seed, (size_t)lda*n, A);
    assert(retval == 0);

    int zerocol = param[PARAM_ZEROCOL].i;
    if (zerocol >= 0 && zerocol < n)
        memset(&A[zerocol*lda], 0, n*sizeof(plasma_complex64_t));

    // Initialize B
    retval = LAPACKE_zlarnv(1, seed, (size_t)ldb*nrhs, B);
    assert(retval == 0);

    plasma_complex64_t *Aref = NULL;
    if (test) {
        Aref = (plasma_complex64_t *)malloc(
            (size_t)lda*n*sizeof(plasma_complex64_t));
        assert(Aref != NULL);
        memcpy(Aref, A, (size_t)lda*n*sizeof(plasma_complex64_t));
    }

    //================================================================
    // Run and time PLASMA
    //================================================================
    plasma_time_t start = omp_get_wtime();
    int plainfo = plasma_zcgesv(n, nrhs, A, lda, ipiv, B, ldb, X, ldx, &ITER);
    plasma_time_t stop = omp_get_wtime();
    plasma_time_t time = stop-start;
    double flops = flops_zgetrf(n, n) + flops_zgetrs(n, nrhs);
    param[PARAM_ITERSV].i = ITER;
    param[PARAM_TIME].d   = time;
    param[PARAM_GFLOPS].d = flops / time / 1e9;

    //================================================================
    // Test results by checking the residual
    //
    //                      || B - AX ||_I
    //                --------------------------- < epsilon
    //                 || A ||_I * || X ||_I * N
    //
    //================================================================
    if (test) {
        if (plainfo == 0) {
            plasma_complex64_t alpha =  1.0;
            plasma_complex64_t beta  = -1.0;

            lapack_int mtrxLayout = LAPACK_COL_MAJOR;
            lapack_int mtrxNorm   = 'I';

            double *work = (double *)malloc(n*sizeof(double));
            assert(work != NULL);

            // Calculate infinite norms of matrices A_ref and X
            double Anorm = LAPACKE_zlange_work(mtrxLayout, mtrxNorm, n, n, Aref,
                                               lda, work);
            double Xnorm = LAPACKE_zlange_work(mtrxLayout, mtrxNorm, n, nrhs, X,
                                               ldx, work);

            // Calculate residual R = A*X-B, store result in B
            cblas_zgemm(CblasColMajor, CblasNoTrans, CblasNoTrans, n, nrhs, n,
                        CBLAS_SADDR(alpha), Aref, lda,
                                            X,    ldx,
                        CBLAS_SADDR(beta),  B,    ldb);

            // Calculate infinite norm of residual matrix R
            double Rnorm = LAPACKE_zlange_work(mtrxLayout, mtrxNorm, n, nrhs, B,
                                               ldb, work);
            // Calculate relative error
            double residual = Rnorm / ( n*Anorm*Xnorm );

            param[PARAM_ERROR].d   = residual;
            param[PARAM_SUCCESS].i = residual < tol;

            free(work);
        }
        else {
            int lapinfo = LAPACKE_zcgesv(
                              LAPACK_COL_MAJOR,
                              n, nrhs, A, lda, ipiv, B, ldb, X, ldx, &ITER);
            if (plainfo == lapinfo) {
                param[PARAM_ERROR].d = 0.0;
                param[PARAM_SUCCESS].i = 1;
            }
            else {
                param[PARAM_ERROR].d = INFINITY;
                param[PARAM_SUCCESS].i = 0;
            }
        }

        free(Aref);
    }

    //================================================================
    // Free arrays
    //================================================================
    free(A); free(ipiv); free(B); free(X);
}

File: fortran_examples/test_zcgesv.f90
!>
!! @file
!!
!!  PLASMA is a software package provided by:
!!  University of Tennessee,  US,
!!  University of Manchester, UK
!!
!! @precisions mixed zc -> ds
!!
!! @brief Tests LU factorization-based iterative refinement routine for
!!  solution of linear system A*X=B, where A is n-by-n matrix, X and B
!!  are n-by-nrhs matrices

      PROGRAM TEST_ZCGESV

      USE, INTRINSIC :: ISO_FORTRAN_ENV
      USE            :: OMP_LIB
      USE            :: PLASMA

      IMPLICIT NONE

      ! Precisions
      integer, parameter :: sp = REAL32
      integer, parameter :: dp = REAL64

      ! Matrices A, X, B
      complex(dp), allocatable :: A(:,:), Aref(:,:), X(:,:), B(:,:)

      ! Matrix dimensions
      integer :: n = 1000, nrhs = 1, lda, ldb, ldx

      ! PLASMA variables
      integer :: nb = 120
      integer :: iter, info

      ! LAPACK variables
      integer :: iSeed(8) = [218, 1910, 3542, 2995, 3607, 27, 3488, 593]

      complex(dp) :: zone = 1.0, zmone = -1.0

      real(dp), allocatable :: work(:)
      real(dp)              :: Anorm, Xnorm, Rnorm, relerr

      ! Pivots
      integer, allocatable :: iPiv(:)

      ! Test variables
      real(dp)          :: tol  = 0.0d+0

      ! Performance variables
      real(dp) :: tstart = 0.0d+0, tstop = 0.0d+0, telapsed = 0.0d+0

      ! External functions
      real(dp), external :: dlamch
      real(dp), external :: zlange


      lda = max(1,n)
      ldb = lda
      ldx = lda

      tol = 50.0d+0 * dlamch('E')

      ! Allocate matrices
      allocate(A(lda,n), Aref(lda,n), X(ldx,nrhs), B(ldb,nrhs), &
               iPiv(lda), work(n))

      !=================================================================
      ! Generate and initialize random matrices A and B
      !=================================================================

      ! Generate random matrices A, B
      call zlarnv(1, iSeed(1:4), lda*n, A)

      ! Save initial matrix A into Aref
      Aref = A

      ! Generate random matrix B
      call zlarnv(1, iSeed(5:8), ldb*nrhs, B)

      ! Initialize solution matrix X
      X = 0.0d+0

      !=================================================================
      ! Initialize PLASMA
      !=================================================================
      call plasma_init(info)

      call plasma_set(PlasmaNb, nb, info)

      !=================================================================
      ! Solve linear system A*X=B
      !=================================================================
      tstart   = omp_get_wtime()

      call plasma_zcgesv(n, nrhs, A, lda, iPiv, B, ldb, X, ldx, &
                         iter, info)

      tstop    = omp_get_wtime()
      telapsed = tstop - tstart

      !=================================================================
      ! Finalize PLASMA
      !=================================================================
      call plasma_finalize(info)

      !=================================================================
      ! Perform linear system A*X=B solution check
      !=================================================================

      ! Calculate infinite norms of matrices Aref and X
      Anorm = zlange('I', n, n,    Aref, lda, work)
      Xnorm = zlange('I', n, nrhs, X,    ldx, work)

      ! Calculate residual R = A*X-B, store result in B
      call zgemm('N', 'N', n, nrhs, n, zone, Aref, lda, X, ldx, &
                  zmone, B, ldb)

      ! Calculate infinite norm of residual matrix R
      Rnorm = zlange('I', n, nrhs, B, ldb, work)

      ! Calculate relative error
      relerr = Rnorm / n / Anorm / Xnorm

      print *, "Time:            ", telapsed
      print *, "Iter. to solve:  ", iter
      print *, "Tolerance:       ", tol
      print *, "Relative error:  ", relerr

      if (relerr < tol) then
        print *, "Solution: CORRECT"
      else
        print *, "Solution: WRONG"
      end if
      write(*,'(a)') ""

      ! Deallocate matrices
      deallocate(A, Aref, X, B, iPiv, work, stat=info)

      END PROGRAM TEST_ZCGESV
